File: backend/services/health_service.py
from repositories.health_repository import HealthRepository


class HealthService:
    """Service layer for health check operations"""

    def __init__(self):
        self.repository = HealthRepository()

    def get_database_health(self) -> dict:
        """
        Check database health and return formatted response

        Returns:
            Dict with service health info including DB status, version, and time

        Raises:
            DatabaseConnectionError if connection fails
            QueryError if query fails
        """
        db_check = self.repository.check_database_connection()

        return {
            "service": "Database Connection Test",
            "database_status": db_check.get("status"),
            "details": {
                "version": db_check.get("server_version"),
                "server_time": db_check.get("server_time")
            }
        }

File: backend-go/services/health_service.go
package services

import (
	"stock-api/models"
	"stock-api/repositories"
)

// HealthService is the service layer for health check operations.
type HealthService struct {
	Repo *repositories.HealthRepository
}

// NewHealthService creates a HealthService backed by the given repository.
func NewHealthService(repo *repositories.HealthRepository) *HealthService {
	return &HealthService{Repo: repo}
}

// GetDatabaseHealth checks the database connection and returns a formatted
// health response. On failure it returns a HealthResponse with a status of
// "error" and the underlying error so the caller can surface the detail.
func (s *HealthService) GetDatabaseHealth() (models.HealthResponse, error) {
	row, err := s.Repo.CheckConnection()
	if err != nil {
		return models.HealthResponse{
			Service:        "Database Connection Test",
			DatabaseStatus: "error",
			Details: map[string]interface{}{
				"error": err.Error(),
			},
		}, err
	}

	return models.HealthResponse{
		Service:        "Database Connection Test",
		DatabaseStatus: "connected",
		Details: map[string]interface{}{
			"version":     row.Version,
			"server_time": row.ServerTime,
		},
	}, nil
}
